packages/adblocker/src/preprocessor.ts

```diff
@@ -0,0 +1,239 @@
+import { StaticDataView, sizeOfUTF8 } from './data-view';
+
+export type EnvKeys =
+  | 'ext_ghostery'
+  | 'ext_abp'
+  | 'ext_ublock'
+  | 'ext_ubol'
+  | 'ext_devbuild'
+  | 'env_chromium'
+  | 'env_edge'
+  | 'env_firefox'
+  | 'env_mobile'
+  | 'env_safari'
+  | 'env_mv3'
+  | 'false'
+  | 'cap_html_filtering'
+  | 'cap_user_stylesheet'
+  | 'adguard'
+  | 'adguard_app_android'
+  | 'adguard_app_ios'
+  | 'adguard_app_mac'
+  | 'adguard_app_windows'
+  | 'adguard_ext_android_cb'
+  | 'adguard_ext_chromium'
+  | 'adguard_ext_edge'
+  | 'adguard_ext_firefox'
+  | 'adguard_ext_opera'
+  | 'adguard_ext_safari'
+  | (string & {});
+
+export class Env extends Map<EnvKeys, boolean> {}
+
+export const enum PreprocessorTypes {
+  INVALID = 0,
+  BEGIF = 1,
+  ELSE = 2,
+  ENDIF = 3,
+}
+
+export function detectPreprocessor(line: string) {
+  // Minimum size of a valid condition should be 6 for something like: "!#if x" or "!#else"
+  if (
+    line.length < 6 ||
+    line.charCodeAt(0) !== 33 /* '!' */ ||
+    line.charCodeAt(1) !== 35 /* '#' */
+  ) {
+    return PreprocessorTypes.INVALID;
+  }
+
+  const command = line.slice(2);
+
+  if (line.startsWith('!#if ')) {
+    return PreprocessorTypes.BEGIF;
+  }
+
+  if (command === 'else') {
+    return PreprocessorTypes.ELSE;
+  }
+
+  if (command === 'endif') {
+    return PreprocessorTypes.ENDIF;
+  }
+
+  return PreprocessorTypes.INVALID;
+}
+
+const tokenizerPattern = /(!|&&|\|\||\(|\)|[a-zA-Z0-9_]+)/g;
+
+const tokenize = (expression: string) => expression.match(tokenizerPattern);
+
+const precedence: Record<string, number> = {
+  '!': 2,
+  '&&': 1,
+  '||': 0,
+};
+
+const isOperator = (token: string) => Object.prototype.hasOwnProperty.call(precedence, token);
+
+/// The parsing is done using the [Shunting yard algorithm](https://en.wikipedia.org/wiki/Shunting_yard_algorithm).
+/// This function takes as input a string expression and an environment Map.
+/// The expression is made of constants (identifiers), logical operators
+/// (&&, ||), negations (!constant) and parentheses.
+///
+/// The environment is a simple Map that associates identifiers to boolean values.
+///
+/// The function should return the result of evaluating the expression using
+/// the values from `environment`. The return value of this function is
+/// either `true` or `false`.
+export const evaluate = (expression: string, env: Env) => {
+  if (!expression.length) {
+    return false;
+  }
+
+  const tokens = tokenize(expression);
+
+  if (!tokens || tokens.length === 0) {
```

Comment:
It can return `null` if there's no match:

```
> ''.match(/(!|&&|\|\||\(|\)|[a-zA-Z0-9_]+)/g)
null
> ''.match(/(!|&&|\|\||\(|\)|[a-zA-Z0-9_]+)/g).length
Uncaught TypeError: Cannot read properties of null (reading 'length')
```